cpp/src/strings/convert/convert_durations.cu

```diff
@@ -0,0 +1,650 @@
+      if (length < item.length) return 1;
+      if (item.item_type == format_char_type::literal) {  // static character we'll just skip;
+        // consume item.length bytes from string
+        ptr += item.length;
+        length -= item.length;
+        continue;
+      }
+
+      // special logic for each specifier
+      int8_t item_length{0};
+      switch (item.value) {
+        case 'd': timeparts[DU_DAY] = str2int(ptr, 11, item_length); break;
+        case '+':
+          if (*ptr == '+') item_length = 1;
+          break;  // skip
+        case 'H': timeparts[DU_HOUR] = str2int(ptr, 2, item_length); break;
+        case 'M': timeparts[DU_MINUTE] = str2int(ptr, 2, item_length); break;
+        case 'S': timeparts[DU_SECOND] = str2int(ptr, 2, item_length); break;
+        case 'u':
+        case 'f':
+          if (*ptr == '.') {
+            auto subsecond_precision = (item.length == -1) ? 9 : item.length - 1;  // +1 is for dot
+            auto subsecond = str2int_fixed(ptr + 1, subsecond_precision, length - 1, item_length);
+            constexpr int64_t powers_of_ten[] = {
+              1L, 10L, 100L, 1000L, 10000L, 100000L, 1000000L, 10000000L, 100000000L, 1000000000L};
+            int64_t nanoseconds =
+              subsecond * powers_of_ten[9 - subsecond_precision];  // normalize to nanoseconds
+            timeparts[DU_SUBSECOND] = nanoseconds;
+            item_length++;
+          }
+          break;
+        default: return 3;
+      }
+      ptr += item_length;
+      length -= item_length;
+    }
+    return 0;
+  }
+
+  __device__ int64_t duration_from_parts(int32_t const* timeparts, type_id units)
+  {
+    int32_t days = timeparts[DU_DAY];
+    if (units == type_id::DURATION_DAYS) return days;
+
+    auto hour     = timeparts[DU_HOUR];
+    auto minute   = timeparts[DU_MINUTE];
+    auto second   = timeparts[DU_SECOND];
+    auto duration = duration_D(days) + simt::std::chrono::hours(hour) +
+                    simt::std::chrono::minutes(minute) + duration_s(second);
+    if (units == type_id::DURATION_SECONDS)
+      return simt::std::chrono::duration_cast<duration_s>(duration).count();
+
+    duration_ns subsecond(timeparts[DU_SUBSECOND]);  // ns
+    if (units == type_id::DURATION_MILLISECONDS) {
+      return simt::std::chrono::duration_cast<duration_ms>(duration + subsecond).count();
+    } else if (units == type_id::DURATION_MICROSECONDS) {
+      return simt::std::chrono::duration_cast<duration_us>(duration + subsecond).count();
+    } else if (units == type_id::DURATION_NANOSECONDS)
+      return simt::std::chrono::duration_cast<duration_ns>(duration + subsecond).count();
+    return simt::std::chrono::duration_cast<duration_ns>(duration + subsecond).count();
+  }
+
+  __device__ T operator()(size_type idx)
+  {
+    if (d_strings.is_null(idx)) return T{0};
+    string_view d_str = d_strings.element<string_view>(idx);
+    if (d_str.empty()) return T{0};
+    //
+    int32_t timeparts[DU_ARRAYSIZE] = {0};
```

Comment:
Using a thread local array like this with an access pattern that isn't known at compile time is going to incur global memory read/writes for accessing the array. Maybe that's unavoidable in the implementation, but just wanted to call awareness to this fact.

Reply:
This happens compulsorily Only if array indices are **not constant** values. In this case, all the indices are enum values (constants). So, compiler will try to put in register memory if possible. If it can't, it will spill to device memory(DRAM).
Even if this is a struct, end result is same. 
aware of this, I will analyse the kernel memusage once to be sure.

Reply:
Ah, that's a good point. I missed that the indices are enum values. I think you're right and you should be safe then.

Reply:
replaced with a struct to make is easy to read.